In [ ]:
import pandas as pd
import numpy as np

from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.base import BaseEstimator, ClassifierMixin
from sklearn.metrics import accuracy_score

# Tensorflow dependencies
import tensorflow as tf
from tensorflow.keras.layers import *
from tensorflow.keras.optimizers import Adam
from tensorflow.keras import backend as K
from tensorflow.keras.callbacks import EarlyStopping, ModelCheckpoint
from tensorflow.keras import regularizers
from tensorflow.keras.losses import SparseCategoricalCrossentropy
from tensorflow.keras.metrics import SparseCategoricalAccuracy
from tensorflow.keras.models import Model

In [ ]:
from google.colab import drive
drive.mount("/content/drive")

data = pd.read_csv('/content/drive/My Drive/secondary_data.csv',sep=';')

Mounted at /content/drive


In [ ]:
print(data)

      class  cap-diameter cap-shape cap-surface cap-color  \
0         p         15.26         x           g         o   
1         p         16.60         x           g         o   
2         p         14.07         x           g         o   
3         p         14.17         f           h         e   
4         p         14.64         x           h         o   
...     ...           ...       ...         ...       ...   
61064     p          1.18         s           s         y   
61065     p          1.27         f           s         y   
61066     p          1.27         s           s         y   
61067     p          1.24         f           s         y   
61068     p          1.17         s           s         y   

      does-bruise-or-bleed gill-attachment gill-spacing gill-color  \
0                        f               e          NaN          w   
1                        f               e          NaN          w   
2                        f               e          NaN  

In [ ]:
data = data = pd.get_dummies(data)

In [ ]:
X = data.drop(columns=['class_p', 'class_e']).values
y = data['class_p'].values

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33)

In [ ]:
#ANN Architecture
class ANNClassifier(BaseEstimator, ClassifierMixin):
    def __init__(self, in_features=10, out_features=2, num_hidden=1,
                 num_neurons=16, regularizer='l2',
                 reg_param=1e-4, batch_norm=False, epochs=100,
                 early_stopping=True, verbose=1):
        self.in_features = in_features
        self.out_features = out_features
        self.num_hidden = num_hidden
        self.num_neurons = num_neurons
        self.regularizer = regularizer
        self.regularizer_func = regularizers.l2 if regularizer=='l2' else regularizers.l1
        self.reg_param = reg_param
        self.batch_norm = batch_norm
        self.epochs = epochs
        self.early_stopping = early_stopping
        self.verbose = verbose

        # Build the ANN
        self.model = ANNClassifier.build_model(in_features, out_features,
                                               self.num_hidden,
                                               self.num_neurons,
                                               self.regularizer_func,
                                               self.reg_param,
                                               self.batch_norm)

    @staticmethod
    def build_model(in_features, out_features,
                    num_hidden, num_neurons,
                    regularizer_func, reg_param, batch_norm):
        inputs = Input(shape=(in_features, ))
        outputs = inputs

        for i in range(num_hidden):
            outputs = Dense(num_neurons, kernel_regularizer=regularizer_func(reg_param), activation='relu')(outputs)
            if(batch_norm):
                outputs = BatchNormalization()(outputs)

        outputs = Dense(out_features, activation='softmax')(outputs)
        model = Model(inputs=inputs, outputs=outputs, name='ANNClassifier')

        return model

    def fit(self, X, Y):
        # Split into training and validating sets
        X_train, X_val, Y_train, Y_val = train_test_split(X, Y, test_size=1/3)

        # Specifies callbacks list
        callbacks = [
            ModelCheckpoint('model.weights.hdf5', save_best_only=True, verbose=self.verbose)
        ]

        # If early stopping is on, append early stopping callback to list
        if(self.early_stopping):
            callbacks.append(EarlyStopping(monitor='val_loss', patience=10, verbose=self.verbose))

        # Configure training process
        adam = Adam(learning_rate=0.001, amsgrad=True)
        loss = SparseCategoricalCrossentropy()
        accuracy = SparseCategoricalAccuracy()

        # Compile the model then train
        self.model.compile(optimizer=adam, loss=loss, metrics=[accuracy])
        history = self.model.fit(X_train, Y_train, validation_data=(X_val, Y_val), epochs=self.epochs,
                       callbacks=callbacks, verbose=self.verbose)

        train_loss = history.history['loss'][-1]
        val_loss = history.history['val_loss'][-1]
        print("Training Loss:", train_loss)
        print("Validation Loss:", val_loss)

        # Calculate training and validation accuracy
        train_accuracy = history.history['sparse_categorical_accuracy'][-1]
        val_accuracy = history.history['val_sparse_categorical_accuracy'][-1]

        # Print or return the metrics
        print("Training Accuracy:", train_accuracy)
        print("Validation Accuracy:", val_accuracy)

        return history

    def predict(self, X):
        predictions = self.model.predict(X)
        predictions = np.argmax(predictions, axis=1)

        return predictions


In [ ]:
exclude = ['class_p','class_e']
features = data.drop(columns=exclude)
in_features = features.shape[1]
out_features = 2

In [ ]:
ann = ANNClassifier(in_features, out_features, epochs=10, verbose=1)
ann.fit(X_train, y_train)

Epoch 1/10
834/853 [============================>.] - ETA: 0s - loss: 0.4253 - sparse_categorical_accuracy: 0.8249
Epoch 1: val_loss improved from inf to 0.21764, saving model to model.weights.hdf5
853/853 [==============================] - 4s 3ms/step - loss: 0.4211 - sparse_categorical_accuracy: 0.8273 - val_loss: 0.2176 - val_sparse_categorical_accuracy: 0.9426
Epoch 2/10
852/853 [============================>.] - ETA: 0s - loss: 0.1489 - sparse_categorical_accuracy: 0.9712
Epoch 2: val_loss improved from 0.21764 to 0.09922, saving model to model.weights.hdf5
853/853 [==============================] - 2s 3ms/step - loss: 0.1488 - sparse_categorical_accuracy: 0.9713 - val_loss: 0.0992 - val_sparse_categorical_accuracy: 0.9868
Epoch 3/10
831/853 [============================>.] - ETA: 0s - loss: 0.0780 - sparse_categorical_accuracy: 0.9914
Epoch 3: val_loss improved from 0.09922 to 0.05936, saving model to model.weights.hdf5
853/853 [==============================] - 2s 3ms/step - los

In [ ]:
param_grid = {
    'num_hidden' : [1, 2],
    'num_neurons' : [8, 16],
    'reg_param' : [2e-4, 2e-3]
}

clf = ANNClassifier(in_features=in_features, out_features=out_features, verbose=1, epochs=3)
gridsearchcv = GridSearchCV(clf, param_grid, cv=3, scoring='accuracy', return_train_score=True, verbose=3)
gridsearchcv.fit(X_train, y_train)
best_params = gridsearchcv.best_params_

Fitting 3 folds for each of 8 candidates, totalling 24 fits
Epoch 1/3
562/569 [============================>.] - ETA: 0s - loss: 0.5056 - sparse_categorical_accuracy: 0.7504
Epoch 1: val_loss improved from inf to 0.34283, saving model to model.weights.hdf5
569/569 [==============================] - 3s 5ms/step - loss: 0.5043 - sparse_categorical_accuracy: 0.7512 - val_loss: 0.3428 - val_sparse_categorical_accuracy: 0.8635
Epoch 2/3
 49/569 [=>............................] - ETA: 1s - loss: 0.3219 - sparse_categorical_accuracy: 0.8903

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


551/569 [============================>.] - ETA: 0s - loss: 0.2635 - sparse_categorical_accuracy: 0.9115
Epoch 2: val_loss improved from 0.34283 to 0.19670, saving model to model.weights.hdf5
569/569 [==============================] - 2s 4ms/step - loss: 0.2614 - sparse_categorical_accuracy: 0.9126 - val_loss: 0.1967 - val_sparse_categorical_accuracy: 0.9495
Epoch 3/3
565/569 [============================>.] - ETA: 0s - loss: 0.1614 - sparse_categorical_accuracy: 0.9604
Epoch 3: val_loss improved from 0.19670 to 0.13383, saving model to model.weights.hdf5
569/569 [==============================] - 2s 3ms/step - loss: 0.1611 - sparse_categorical_accuracy: 0.9606 - val_loss: 0.1338 - val_sparse_categorical_accuracy: 0.9718
Training Loss: 0.16108502447605133
Validation Loss: 0.13382774591445923
Training Accuracy: 0.9605697393417358
Validation Accuracy: 0.9718464612960815
853/853 [==============================] - 1s 2ms/step
[CV 1/3] END num_hidden=1, num_neurons=8, reg_param=0.0002;, scor

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


552/569 [============================>.] - ETA: 0s - loss: 0.2630 - sparse_categorical_accuracy: 0.9168
Epoch 2: val_loss improved from 0.34193 to 0.20066, saving model to model.weights.hdf5
569/569 [==============================] - 2s 4ms/step - loss: 0.2617 - sparse_categorical_accuracy: 0.9168 - val_loss: 0.2007 - val_sparse_categorical_accuracy: 0.9426
Epoch 3/3
562/569 [============================>.] - ETA: 0s - loss: 0.1581 - sparse_categorical_accuracy: 0.9610
Epoch 3: val_loss improved from 0.20066 to 0.12498, saving model to model.weights.hdf5
569/569 [==============================] - 2s 3ms/step - loss: 0.1577 - sparse_categorical_accuracy: 0.9613 - val_loss: 0.1250 - val_sparse_categorical_accuracy: 0.9781
Training Loss: 0.15769754350185394
Validation Loss: 0.12497959285974503
Training Accuracy: 0.9612846374511719
Validation Accuracy: 0.9781150221824646
853/853 [==============================] - 1s 1ms/step
[CV 2/3] END num_hidden=1, num_neurons=8, reg_param=0.0002;, scor

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


563/569 [============================>.] - ETA: 0s - loss: 0.2555 - sparse_categorical_accuracy: 0.9124
Epoch 2: val_loss improved from 0.33119 to 0.19449, saving model to model.weights.hdf5
569/569 [==============================] - 2s 3ms/step - loss: 0.2546 - sparse_categorical_accuracy: 0.9129 - val_loss: 0.1945 - val_sparse_categorical_accuracy: 0.9433
Epoch 3/3
562/569 [============================>.] - ETA: 0s - loss: 0.1598 - sparse_categorical_accuracy: 0.9588
Epoch 3: val_loss improved from 0.19449 to 0.12726, saving model to model.weights.hdf5
569/569 [==============================] - 2s 3ms/step - loss: 0.1594 - sparse_categorical_accuracy: 0.9590 - val_loss: 0.1273 - val_sparse_categorical_accuracy: 0.9714
Training Loss: 0.1594463437795639
Validation Loss: 0.12725938856601715
Training Accuracy: 0.9589771628379822
Validation Accuracy: 0.9714065790176392
853/853 [==============================] - 1s 1ms/step
[CV 3/3] END num_hidden=1, num_neurons=8, reg_param=0.0002;, score

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


553/569 [============================>.] - ETA: 0s - loss: 0.2724 - sparse_categorical_accuracy: 0.8993
Epoch 2: val_loss improved from 0.34308 to 0.21681, saving model to model.weights.hdf5
569/569 [==============================] - 2s 4ms/step - loss: 0.2702 - sparse_categorical_accuracy: 0.9007 - val_loss: 0.2168 - val_sparse_categorical_accuracy: 0.9235
Epoch 3/3
559/569 [============================>.] - ETA: 0s - loss: 0.1800 - sparse_categorical_accuracy: 0.9420
Epoch 3: val_loss improved from 0.21681 to 0.14632, saving model to model.weights.hdf5
569/569 [==============================] - 2s 4ms/step - loss: 0.1798 - sparse_categorical_accuracy: 0.9422 - val_loss: 0.1463 - val_sparse_categorical_accuracy: 0.9614
Training Loss: 0.17980319261550903
Validation Loss: 0.14631564915180206
Training Accuracy: 0.9422019124031067
Validation Accuracy: 0.9613988995552063
853/853 [==============================] - 2s 2ms/step
[CV 1/3] END num_hidden=1, num_neurons=8, reg_param=0.002;, score

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


562/569 [============================>.] - ETA: 0s - loss: 0.2939 - sparse_categorical_accuracy: 0.8884
Epoch 2: val_loss improved from 0.36410 to 0.23318, saving model to model.weights.hdf5
569/569 [==============================] - 2s 3ms/step - loss: 0.2932 - sparse_categorical_accuracy: 0.8885 - val_loss: 0.2332 - val_sparse_categorical_accuracy: 0.9210
Epoch 3/3
547/569 [===========================>..] - ETA: 0s - loss: 0.1949 - sparse_categorical_accuracy: 0.9437
Epoch 3: val_loss improved from 0.23318 to 0.15887, saving model to model.weights.hdf5
569/569 [==============================] - 2s 3ms/step - loss: 0.1939 - sparse_categorical_accuracy: 0.9440 - val_loss: 0.1589 - val_sparse_categorical_accuracy: 0.9593
Training Loss: 0.193903848528862
Validation Loss: 0.15886639058589935
Training Accuracy: 0.9439617395401001
Validation Accuracy: 0.9593093395233154
853/853 [==============================] - 1s 1ms/step
[CV 2/3] END num_hidden=1, num_neurons=8, reg_param=0.002;, score=(

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


566/569 [============================>.] - ETA: 0s - loss: 0.2828 - sparse_categorical_accuracy: 0.9021
Epoch 2: val_loss improved from 0.35474 to 0.22159, saving model to model.weights.hdf5
569/569 [==============================] - 2s 3ms/step - loss: 0.2827 - sparse_categorical_accuracy: 0.9022 - val_loss: 0.2216 - val_sparse_categorical_accuracy: 0.9337
Epoch 3/3
556/569 [============================>.] - ETA: 0s - loss: 0.1853 - sparse_categorical_accuracy: 0.9501
Epoch 3: val_loss improved from 0.22159 to 0.15258, saving model to model.weights.hdf5
569/569 [==============================] - 2s 4ms/step - loss: 0.1845 - sparse_categorical_accuracy: 0.9506 - val_loss: 0.1526 - val_sparse_categorical_accuracy: 0.9652
Training Loss: 0.18448373675346375
Validation Loss: 0.15257763862609863
Training Accuracy: 0.9505636692047119
Validation Accuracy: 0.9652479887008667
853/853 [==============================] - 1s 2ms/step
[CV 3/3] END num_hidden=1, num_neurons=8, reg_param=0.002;, score

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


556/569 [============================>.] - ETA: 0s - loss: 0.2490 - sparse_categorical_accuracy: 0.9184
Epoch 2: val_loss improved from 0.32959 to 0.20046, saving model to model.weights.hdf5
569/569 [==============================] - 2s 3ms/step - loss: 0.2485 - sparse_categorical_accuracy: 0.9182 - val_loss: 0.2005 - val_sparse_categorical_accuracy: 0.9395
Epoch 3/3
551/569 [============================>.] - ETA: 0s - loss: 0.1524 - sparse_categorical_accuracy: 0.9605
Epoch 3: val_loss improved from 0.20046 to 0.12157, saving model to model.weights.hdf5
569/569 [==============================] - 2s 4ms/step - loss: 0.1513 - sparse_categorical_accuracy: 0.9612 - val_loss: 0.1216 - val_sparse_categorical_accuracy: 0.9776
Training Loss: 0.15128619968891144
Validation Loss: 0.12156840413808823
Training Accuracy: 0.9611746668815613
Validation Accuracy: 0.9775651693344116
853/853 [==============================] - 1s 1ms/step
[CV 1/3] END num_hidden=1, num_neurons=16, reg_param=0.0002;, sco

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


547/569 [===========================>..] - ETA: 0s - loss: 0.2351 - sparse_categorical_accuracy: 0.9309
Epoch 2: val_loss improved from 0.32689 to 0.15693, saving model to model.weights.hdf5
569/569 [==============================] - 2s 3ms/step - loss: 0.2326 - sparse_categorical_accuracy: 0.9325 - val_loss: 0.1569 - val_sparse_categorical_accuracy: 0.9695
Epoch 3/3
553/569 [============================>.] - ETA: 0s - loss: 0.1205 - sparse_categorical_accuracy: 0.9823
Epoch 3: val_loss improved from 0.15693 to 0.09293, saving model to model.weights.hdf5
569/569 [==============================] - 2s 4ms/step - loss: 0.1195 - sparse_categorical_accuracy: 0.9827 - val_loss: 0.0929 - val_sparse_categorical_accuracy: 0.9883
Training Loss: 0.11948184669017792
Validation Loss: 0.09292588382959366
Training Accuracy: 0.9827320575714111
Validation Accuracy: 0.9883427023887634
853/853 [==============================] - 1s 1ms/step
[CV 2/3] END num_hidden=1, num_neurons=16, reg_param=0.0002;, sco

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


545/569 [===========================>..] - ETA: 0s - loss: 0.2928 - sparse_categorical_accuracy: 0.8920
Epoch 2: val_loss improved from 0.35996 to 0.22862, saving model to model.weights.hdf5
569/569 [==============================] - 2s 3ms/step - loss: 0.2908 - sparse_categorical_accuracy: 0.8930 - val_loss: 0.2286 - val_sparse_categorical_accuracy: 0.9307
Epoch 3/3
558/569 [============================>.] - ETA: 0s - loss: 0.1862 - sparse_categorical_accuracy: 0.9462
Epoch 3: val_loss improved from 0.22862 to 0.14825, saving model to model.weights.hdf5
569/569 [==============================] - 2s 3ms/step - loss: 0.1854 - sparse_categorical_accuracy: 0.9465 - val_loss: 0.1482 - val_sparse_categorical_accuracy: 0.9660
Training Loss: 0.18540944159030914
Validation Loss: 0.14824552834033966
Training Accuracy: 0.9464943408966064
Validation Accuracy: 0.9660178422927856
853/853 [==============================] - 2s 2ms/step
[CV 3/3] END num_hidden=1, num_neurons=16, reg_param=0.0002;, sco

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


562/569 [============================>.] - ETA: 0s - loss: 0.2757 - sparse_categorical_accuracy: 0.9001
Epoch 2: val_loss improved from 0.36389 to 0.21674, saving model to model.weights.hdf5
569/569 [==============================] - 2s 3ms/step - loss: 0.2751 - sparse_categorical_accuracy: 0.9004 - val_loss: 0.2167 - val_sparse_categorical_accuracy: 0.9298
Epoch 3/3
550/569 [===========================>..] - ETA: 0s - loss: 0.1768 - sparse_categorical_accuracy: 0.9476
Epoch 3: val_loss improved from 0.21674 to 0.14918, saving model to model.weights.hdf5
569/569 [==============================] - 2s 3ms/step - loss: 0.1757 - sparse_categorical_accuracy: 0.9483 - val_loss: 0.1492 - val_sparse_categorical_accuracy: 0.9593
Training Loss: 0.17572954297065735
Validation Loss: 0.14917674660682678
Training Accuracy: 0.9483062028884888
Validation Accuracy: 0.9593093395233154
853/853 [==============================] - 1s 1ms/step
[CV 1/3] END num_hidden=1, num_neurons=16, reg_param=0.002;, scor

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


561/569 [============================>.] - ETA: 0s - loss: 0.2569 - sparse_categorical_accuracy: 0.9120
Epoch 2: val_loss improved from 0.34645 to 0.19892, saving model to model.weights.hdf5
569/569 [==============================] - 2s 3ms/step - loss: 0.2560 - sparse_categorical_accuracy: 0.9123 - val_loss: 0.1989 - val_sparse_categorical_accuracy: 0.9372
Epoch 3/3
565/569 [============================>.] - ETA: 0s - loss: 0.1568 - sparse_categorical_accuracy: 0.9570
Epoch 3: val_loss improved from 0.19892 to 0.13272, saving model to model.weights.hdf5
569/569 [==============================] - 2s 3ms/step - loss: 0.1567 - sparse_categorical_accuracy: 0.9571 - val_loss: 0.1327 - val_sparse_categorical_accuracy: 0.9728
Training Loss: 0.15665973722934723
Validation Loss: 0.13272187113761902
Training Accuracy: 0.9571051597595215
Validation Accuracy: 0.9728362560272217
853/853 [==============================] - 1s 1ms/step
[CV 2/3] END num_hidden=1, num_neurons=16, reg_param=0.002;, scor

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


555/569 [============================>.] - ETA: 0s - loss: 0.2699 - sparse_categorical_accuracy: 0.9055
Epoch 2: val_loss improved from 0.34145 to 0.20935, saving model to model.weights.hdf5
569/569 [==============================] - 2s 4ms/step - loss: 0.2686 - sparse_categorical_accuracy: 0.9063 - val_loss: 0.2093 - val_sparse_categorical_accuracy: 0.9323
Epoch 3/3
551/569 [============================>.] - ETA: 0s - loss: 0.1682 - sparse_categorical_accuracy: 0.9549
Epoch 3: val_loss improved from 0.20935 to 0.13468, saving model to model.weights.hdf5
569/569 [==============================] - 3s 5ms/step - loss: 0.1679 - sparse_categorical_accuracy: 0.9547 - val_loss: 0.1347 - val_sparse_categorical_accuracy: 0.9680
Training Loss: 0.1678769439458847
Validation Loss: 0.13467970490455627
Training Accuracy: 0.9546879529953003
Validation Accuracy: 0.9679973721504211
853/853 [==============================] - 1s 2ms/step
[CV 3/3] END num_hidden=1, num_neurons=16, reg_param=0.002;, score

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


552/569 [============================>.] - ETA: 0s - loss: 0.2419 - sparse_categorical_accuracy: 0.9147
Epoch 2: val_loss improved from 0.31815 to 0.18070, saving model to model.weights.hdf5
569/569 [==============================] - 2s 3ms/step - loss: 0.2409 - sparse_categorical_accuracy: 0.9149 - val_loss: 0.1807 - val_sparse_categorical_accuracy: 0.9462
Epoch 3/3
553/569 [============================>.] - ETA: 0s - loss: 0.1402 - sparse_categorical_accuracy: 0.9658
Epoch 3: val_loss improved from 0.18070 to 0.11102, saving model to model.weights.hdf5
569/569 [==============================] - 2s 3ms/step - loss: 0.1393 - sparse_categorical_accuracy: 0.9662 - val_loss: 0.1110 - val_sparse_categorical_accuracy: 0.9783
Training Loss: 0.1393297016620636
Validation Loss: 0.11102259904146194
Training Accuracy: 0.966234028339386
Validation Accuracy: 0.9783349633216858
853/853 [==============================] - 1s 1ms/step
[CV 1/3] END num_hidden=2, num_neurons=8, reg_param=0.0002;, score=

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


563/569 [============================>.] - ETA: 0s - loss: 0.2593 - sparse_categorical_accuracy: 0.9121
Epoch 2: val_loss improved from 0.33340 to 0.20594, saving model to model.weights.hdf5
569/569 [==============================] - 2s 4ms/step - loss: 0.2589 - sparse_categorical_accuracy: 0.9122 - val_loss: 0.2059 - val_sparse_categorical_accuracy: 0.9359
Epoch 3/3
563/569 [============================>.] - ETA: 0s - loss: 0.1678 - sparse_categorical_accuracy: 0.9580
Epoch 3: val_loss improved from 0.20594 to 0.14350, saving model to model.weights.hdf5
569/569 [==============================] - 3s 6ms/step - loss: 0.1678 - sparse_categorical_accuracy: 0.9580 - val_loss: 0.1435 - val_sparse_categorical_accuracy: 0.9621
Training Loss: 0.16781890392303467
Validation Loss: 0.14349548518657684
Training Accuracy: 0.9580400586128235
Validation Accuracy: 0.9620587229728699
853/853 [==============================] - 1s 2ms/step
[CV 2/3] END num_hidden=2, num_neurons=8, reg_param=0.0002;, scor

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


568/569 [============================>.] - ETA: 0s - loss: 0.2582 - sparse_categorical_accuracy: 0.9146
Epoch 2: val_loss improved from 0.35079 to 0.19978, saving model to model.weights.hdf5
569/569 [==============================] - 2s 3ms/step - loss: 0.2582 - sparse_categorical_accuracy: 0.9145 - val_loss: 0.1998 - val_sparse_categorical_accuracy: 0.9383
Epoch 3/3
569/569 [==============================] - ETA: 0s - loss: 0.1554 - sparse_categorical_accuracy: 0.9586
Epoch 3: val_loss improved from 0.19978 to 0.13365, saving model to model.weights.hdf5
569/569 [==============================] - 3s 5ms/step - loss: 0.1554 - sparse_categorical_accuracy: 0.9586 - val_loss: 0.1336 - val_sparse_categorical_accuracy: 0.9685
Training Loss: 0.1554446965456009
Validation Loss: 0.1336456686258316
Training Accuracy: 0.9585922360420227
Validation Accuracy: 0.9685472249984741
853/853 [==============================] - 1s 1ms/step
[CV 3/3] END num_hidden=2, num_neurons=8, reg_param=0.0002;, score=

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


568/569 [============================>.] - ETA: 0s - loss: 0.2667 - sparse_categorical_accuracy: 0.9116
Epoch 2: val_loss improved from 0.34179 to 0.20640, saving model to model.weights.hdf5
569/569 [==============================] - 2s 4ms/step - loss: 0.2667 - sparse_categorical_accuracy: 0.9116 - val_loss: 0.2064 - val_sparse_categorical_accuracy: 0.9380
Epoch 3/3
550/569 [===========================>..] - ETA: 0s - loss: 0.1703 - sparse_categorical_accuracy: 0.9559
Epoch 3: val_loss improved from 0.20640 to 0.13881, saving model to model.weights.hdf5
569/569 [==============================] - 2s 4ms/step - loss: 0.1692 - sparse_categorical_accuracy: 0.9564 - val_loss: 0.1388 - val_sparse_categorical_accuracy: 0.9661
Training Loss: 0.16923968493938446
Validation Loss: 0.13880635797977448
Training Accuracy: 0.9563902616500854
Validation Accuracy: 0.9661278128623962
853/853 [==============================] - 1s 1ms/step
[CV 1/3] END num_hidden=2, num_neurons=8, reg_param=0.002;, score

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


561/569 [============================>.] - ETA: 0s - loss: 0.2506 - sparse_categorical_accuracy: 0.9161
Epoch 2: val_loss improved from 0.33155 to 0.18897, saving model to model.weights.hdf5
569/569 [==============================] - 2s 3ms/step - loss: 0.2495 - sparse_categorical_accuracy: 0.9166 - val_loss: 0.1890 - val_sparse_categorical_accuracy: 0.9507
Epoch 3/3
566/569 [============================>.] - ETA: 0s - loss: 0.1483 - sparse_categorical_accuracy: 0.9673
Epoch 3: val_loss improved from 0.18897 to 0.11826, saving model to model.weights.hdf5
569/569 [==============================] - 2s 4ms/step - loss: 0.1483 - sparse_categorical_accuracy: 0.9673 - val_loss: 0.1183 - val_sparse_categorical_accuracy: 0.9810
Training Loss: 0.14826324582099915
Validation Loss: 0.11825897544622421
Training Accuracy: 0.9673339128494263
Validation Accuracy: 0.9809743762016296
853/853 [==============================] - 1s 1ms/step
[CV 2/3] END num_hidden=2, num_neurons=8, reg_param=0.002;, score

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


563/569 [============================>.] - ETA: 0s - loss: 0.3432 - sparse_categorical_accuracy: 0.8658
Epoch 2: val_loss improved from 0.40248 to 0.28152, saving model to model.weights.hdf5
569/569 [==============================] - 2s 3ms/step - loss: 0.3425 - sparse_categorical_accuracy: 0.8662 - val_loss: 0.2815 - val_sparse_categorical_accuracy: 0.8945
Epoch 3/3
555/569 [============================>.] - ETA: 0s - loss: 0.2272 - sparse_categorical_accuracy: 0.9271
Epoch 3: val_loss improved from 0.28152 to 0.17956, saving model to model.weights.hdf5
569/569 [==============================] - 2s 3ms/step - loss: 0.2258 - sparse_categorical_accuracy: 0.9279 - val_loss: 0.1796 - val_sparse_categorical_accuracy: 0.9515
Training Loss: 0.22580616176128387
Validation Loss: 0.17955821752548218
Training Accuracy: 0.9279076457023621
Validation Accuracy: 0.9515011310577393
853/853 [==============================] - 2s 2ms/step
[CV 3/3] END num_hidden=2, num_neurons=8, reg_param=0.002;, score

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


566/569 [============================>.] - ETA: 0s - loss: 0.2857 - sparse_categorical_accuracy: 0.8971
Epoch 2: val_loss improved from 0.36304 to 0.23234, saving model to model.weights.hdf5
569/569 [==============================] - 2s 3ms/step - loss: 0.2858 - sparse_categorical_accuracy: 0.8971 - val_loss: 0.2323 - val_sparse_categorical_accuracy: 0.9219
Epoch 3/3
549/569 [===========================>..] - ETA: 0s - loss: 0.1822 - sparse_categorical_accuracy: 0.9508
Epoch 3: val_loss improved from 0.23234 to 0.15081, saving model to model.weights.hdf5
569/569 [==============================] - 2s 3ms/step - loss: 0.1810 - sparse_categorical_accuracy: 0.9511 - val_loss: 0.1508 - val_sparse_categorical_accuracy: 0.9632
Training Loss: 0.18102753162384033
Validation Loss: 0.1508110612630844
Training Accuracy: 0.9510558843612671
Validation Accuracy: 0.9631584882736206
853/853 [==============================] - 1s 1ms/step
[CV 1/3] END num_hidden=2, num_neurons=16, reg_param=0.0002;, scor

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


561/569 [============================>.] - ETA: 0s - loss: 0.2886 - sparse_categorical_accuracy: 0.8964
Epoch 2: val_loss improved from 0.35742 to 0.24210, saving model to model.weights.hdf5
569/569 [==============================] - 3s 4ms/step - loss: 0.2876 - sparse_categorical_accuracy: 0.8971 - val_loss: 0.2421 - val_sparse_categorical_accuracy: 0.9139
Epoch 3/3
554/569 [============================>.] - ETA: 0s - loss: 0.2000 - sparse_categorical_accuracy: 0.9346
Epoch 3: val_loss improved from 0.24210 to 0.17621, saving model to model.weights.hdf5
569/569 [==============================] - 2s 3ms/step - loss: 0.1992 - sparse_categorical_accuracy: 0.9350 - val_loss: 0.1762 - val_sparse_categorical_accuracy: 0.9376
Training Loss: 0.19918474555015564
Validation Loss: 0.17620691657066345
Training Accuracy: 0.9349977970123291
Validation Accuracy: 0.937644362449646
853/853 [==============================] - 1s 1ms/step
[CV 2/3] END num_hidden=2, num_neurons=16, reg_param=0.0002;, scor

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


556/569 [============================>.] - ETA: 0s - loss: 0.2513 - sparse_categorical_accuracy: 0.9192
Epoch 2: val_loss improved from 0.34457 to 0.18926, saving model to model.weights.hdf5
569/569 [==============================] - 2s 3ms/step - loss: 0.2497 - sparse_categorical_accuracy: 0.9200 - val_loss: 0.1893 - val_sparse_categorical_accuracy: 0.9486
Epoch 3/3
563/569 [============================>.] - ETA: 0s - loss: 0.1461 - sparse_categorical_accuracy: 0.9668
Epoch 3: val_loss improved from 0.18926 to 0.12075, saving model to model.weights.hdf5
569/569 [==============================] - 3s 5ms/step - loss: 0.1458 - sparse_categorical_accuracy: 0.9669 - val_loss: 0.1208 - val_sparse_categorical_accuracy: 0.9739
Training Loss: 0.14579342305660248
Validation Loss: 0.12075060606002808
Training Accuracy: 0.9668958187103271
Validation Accuracy: 0.9739360213279724
853/853 [==============================] - 1s 1ms/step
[CV 3/3] END num_hidden=2, num_neurons=16, reg_param=0.0002;, sco

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


565/569 [============================>.] - ETA: 0s - loss: 0.2168 - sparse_categorical_accuracy: 0.9386
Epoch 2: val_loss improved from 0.30940 to 0.15105, saving model to model.weights.hdf5
569/569 [==============================] - 2s 4ms/step - loss: 0.2166 - sparse_categorical_accuracy: 0.9387 - val_loss: 0.1511 - val_sparse_categorical_accuracy: 0.9698
Epoch 3/3
568/569 [============================>.] - ETA: 0s - loss: 0.1165 - sparse_categorical_accuracy: 0.9816
Epoch 3: val_loss improved from 0.15105 to 0.08973, saving model to model.weights.hdf5
569/569 [==============================] - 2s 4ms/step - loss: 0.1165 - sparse_categorical_accuracy: 0.9816 - val_loss: 0.0897 - val_sparse_categorical_accuracy: 0.9882
Training Loss: 0.11648157984018326
Validation Loss: 0.08972704410552979
Training Accuracy: 0.9815772175788879
Validation Accuracy: 0.9882327318191528
853/853 [==============================] - 1s 2ms/step
[CV 1/3] END num_hidden=2, num_neurons=16, reg_param=0.002;, scor

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


565/569 [============================>.] - ETA: 0s - loss: 0.2455 - sparse_categorical_accuracy: 0.9199
Epoch 2: val_loss improved from 0.32797 to 0.18457, saving model to model.weights.hdf5
569/569 [==============================] - 2s 3ms/step - loss: 0.2452 - sparse_categorical_accuracy: 0.9199 - val_loss: 0.1846 - val_sparse_categorical_accuracy: 0.9424
Epoch 3/3
557/569 [============================>.] - ETA: 0s - loss: 0.1461 - sparse_categorical_accuracy: 0.9652
Epoch 3: val_loss improved from 0.18457 to 0.11908, saving model to model.weights.hdf5
569/569 [==============================] - 2s 3ms/step - loss: 0.1456 - sparse_categorical_accuracy: 0.9653 - val_loss: 0.1191 - val_sparse_categorical_accuracy: 0.9781
Training Loss: 0.14555658400058746
Validation Loss: 0.11908017098903656
Training Accuracy: 0.9652991890907288
Validation Accuracy: 0.9781150221824646
853/853 [==============================] - 1s 1ms/step
[CV 2/3] END num_hidden=2, num_neurons=16, reg_param=0.002;, scor

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


561/569 [============================>.] - ETA: 0s - loss: 0.2483 - sparse_categorical_accuracy: 0.9206
Epoch 2: val_loss improved from 0.32907 to 0.18622, saving model to model.weights.hdf5
569/569 [==============================] - 2s 3ms/step - loss: 0.2476 - sparse_categorical_accuracy: 0.9208 - val_loss: 0.1862 - val_sparse_categorical_accuracy: 0.9473
Epoch 3/3
564/569 [============================>.] - ETA: 0s - loss: 0.1431 - sparse_categorical_accuracy: 0.9638
Epoch 3: val_loss improved from 0.18622 to 0.11481, saving model to model.weights.hdf5
569/569 [==============================] - 3s 4ms/step - loss: 0.1428 - sparse_categorical_accuracy: 0.9639 - val_loss: 0.1148 - val_sparse_categorical_accuracy: 0.9726
Training Loss: 0.1427573263645172
Validation Loss: 0.11481349915266037
Training Accuracy: 0.9639263153076172
Validation Accuracy: 0.9726163148880005
853/853 [==============================] - 1s 1ms/step
[CV 3/3] END num_hidden=2, num_neurons=16, reg_param=0.002;, score

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


847/853 [============================>.] - ETA: 0s - loss: 0.1551 - sparse_categorical_accuracy: 0.9891
Epoch 2: val_loss improved from 0.21065 to 0.12298, saving model to model.weights.hdf5
853/853 [==============================] - 3s 4ms/step - loss: 0.1549 - sparse_categorical_accuracy: 0.9891 - val_loss: 0.1230 - val_sparse_categorical_accuracy: 0.9970
Epoch 3/3
847/853 [============================>.] - ETA: 0s - loss: 0.1119 - sparse_categorical_accuracy: 0.9973
Epoch 3: val_loss improved from 0.12298 to 0.10213, saving model to model.weights.hdf5
853/853 [==============================] - 3s 4ms/step - loss: 0.1118 - sparse_categorical_accuracy: 0.9973 - val_loss: 0.1021 - val_sparse_categorical_accuracy: 0.9984
Training Loss: 0.11184718459844589
Validation Loss: 0.10212936252355576
Training Accuracy: 0.997323751449585
Validation Accuracy: 0.9983869791030884


In [ ]:
print('Best model parameters : ', best_params)

Best model parameters :  {'num_hidden': 2, 'num_neurons': 16, 'reg_param': 0.002}


In [ ]:
clf =  ANNClassifier(in_features=in_features, out_features=out_features, verbose=1, epochs=5, **best_params)
clf.fit(X_train, y_train)

Epoch 1/5
834/853 [============================>.] - ETA: 0s - loss: 0.4301 - sparse_categorical_accuracy: 0.8398
Epoch 1: val_loss improved from inf to 0.23085, saving model to model.weights.hdf5
853/853 [==============================] - 3s 3ms/step - loss: 0.4261 - sparse_categorical_accuracy: 0.8420 - val_loss: 0.2308 - val_sparse_categorical_accuracy: 0.9559
Epoch 2/5
849/853 [============================>.] - ETA: 0s - loss: 0.1760 - sparse_categorical_accuracy: 0.9771
Epoch 2: val_loss improved from 0.23085 to 0.14003, saving model to model.weights.hdf5
853/853 [==============================] - 3s 3ms/step - loss: 0.1759 - sparse_categorical_accuracy: 0.9771 - val_loss: 0.1400 - val_sparse_categorical_accuracy: 0.9911
Epoch 3/5
844/853 [============================>.] - ETA: 0s - loss: 0.1265 - sparse_categorical_accuracy: 0.9927
Epoch 3: val_loss improved from 0.14003 to 0.11392, saving model to model.weights.hdf5
853/853 [==============================] - 2s 3ms/step - loss: 

In [ ]:
predictions = clf.predict(X_test)
accuracy = accuracy_score(y_test, predictions)

print('Accuracy on testing set : ', accuracy)

630/630 [==============================] - 1s 2ms/step
Accuracy on testing set :  0.9974693594005856


Before Tuning

*   Training Loss: 0.022285955026745796
*   Validation Loss: 0.021776264533400536
*   Training Accuracy: 0.9998900294303894
*   Validation Accuracy: 0.9996334314346313

Best model parameters :  {'num_hidden': 2, 'num_neurons': 16, 'reg_param': 0.002}

After Tuning:

*   Training Loss: 0.09437078982591629
*   Validation Loss: 0.09008152782917023
*   Training Accuracy: 0.9971771240234375
*   Validation Accuracy: 0.9980937242507935



Accuracy on testing set :  0.9974693594005856